ocaml/xapi/xapi_vdi.ml

```diff
@@ -690,7 +690,12 @@ let set_snapshot_time ~__context ~self ~value =
   Db.VDI.set_snapshot_time ~__context ~self ~value
 
 let set_metadata_of_pool ~__context ~self ~value =
-  Db.VDI.set_metadata_of_pool ~__context ~self ~value
+  let vdi_type = Db.VDI.get_type ~__context ~self in
+  match vdi_type with
+  | `metadata -> Db.VDI.set_metadata_of_pool ~__context ~self ~value
+  | _ ->
+    error "VDI.set_metadata_of_pool: called with a VDI that does not have type metadata (at %s)" __LOC__;
```

Comment:
Hmm yeah, I was actually wondering whether we need this error message in addition to the exception, whether it provides any additional useful information. And in general, I don't know whether we should follow this pattern of logging error + raising an exception, or just raise an exception. The reason I didn't add it is that I thought that it's sufficient to have it in the error, but I'll add it if it's useful to have the VDI ref here too, in addition to the XenAPI error.